packages/monaco/src/browser/monaco-editor-provider.ts

```diff
@@ -276,6 +278,7 @@ export class MonacoEditorProvider {
             }
         }));
         toDispose.push(editor.onLanguageChanged(() => this.updateMonacoEditorOptions(editor)));
+        editor.document.onWillSaveModel(event => event.waitUntil(this.removeTrailingWhiteSpace(editor, event)));
         editor.document.onWillSaveModel(event => event.waitUntil(this.formatOnSave(editor, event)));
```

Comment:
Should we re-use the same listener?

Reply:
I wondered that as well. One question is what the behavior of the Monaco editor is when it processes these hooks. As the code is structured, the calls to perform the two actions here (formatting and whitespace removal) will be made concurrently. Does Monaco implement some transaction protection to ensure that the edits aren't overwriting each other, or does that have to be managed by the caller?

Reply:
I've moved this into the same listener, running format first, then whitespace trimming. Better safe than sorry, even if Monaco is smart enough to handle the two requests intelligently.